ssz_serialization/merkleization.nim

```diff
@@ -635,6 +721,248 @@ func hashTreeRootAux[T](x: T): Digest =
+    const
+      totalChunks = maxChunksCount(T, x.maxLen)
+      firstChunkIndex = nextPow2(totalChunks.uint64)
+      chunkLayer = log2trunc(firstChunkIndex)
+    var i = slice.a
+    while i <= slice.b:
+      let
+        index = indexAt(i)
+        indexLayer = log2trunc(index)
+      if index == 1.GeneralizedIndex:
+        let contentsHash = bitListHashTreeRoot(totalChunks, BitSeq x)
+        rootAt(i) = mixInLength(contentsHash, x.len)
+        inc i
+      elif index == 3.GeneralizedIndex:
+        rootAt(i) = hashTreeRootAux(toSszType(x.len.uint64))
+        inc i
+      elif index == 2.GeneralizedIndex:
+        rootAt(i) = bitListHashTreeRoot(totalChunks, BitSeq x)
+        inc i
+      elif (index shr (indexLayer - 1)) == 2.GeneralizedIndex:
+        let
+          atLayer = atLayer + 1
+          index = indexAt(i)
+          indexLayer = log2trunc(index)
+        if indexLayer <= chunkLayer:
+          let chunks = chunksForIndex(index)
+          rootAt(i) = bitListHashTreeRoot(totalChunks, BitSeq x,
+                                          chunks, indexLayer)
+          inc i
+        else: return unsupportedIndex
+      else: return unsupportedIndex
+  elif T is array:
+    type E = ElemType(T)
+    when E is BasicType and sizeof(T) <= sizeof(roots[0].data):
+      for i in slice:
+        if indexAt(i) != 1.GeneralizedIndex: return unsupportedIndex
+        rootAt(i) = hashTreeRootAux(x)
+    else:
+      trs "FIXED TYPE; USE CHUNK STREAM"
+      const
+        totalChunks = maxChunksCount(T, x.len)
+        firstChunkIndex = nextPow2(totalChunks.uint64)
+        chunkLayer = log2trunc(firstChunkIndex)
+      var i = slice.a
+      while i <= slice.b:
+        let
+          index = indexAt(i)
+          indexLayer = log2trunc(index)
+        if index == 1.GeneralizedIndex:
+          rootAt(i) = chunkedHashTreeRoot(totalChunks, x)
+          inc i
+        elif indexLayer <= chunkLayer:
+          let chunks = chunksForIndex(index)
+          rootAt(i) = chunkedHashTreeRoot(totalChunks, x,
+                                          chunks, indexLayer)
+          inc i
+        else:
+          when ElemType(typeof(x)) is BasicType: return unsupportedIndex
+          else:
+            let chunk = chunkContainingIndex(index)
+            if chunk >= x.len: return unsupportedIndex
+            var j = i + 1
+            while j <= slice.b and
+                chunkContainingIndex(indexAt(j)) == chunk:
+              inc j
+            ? hash_tree_root_multi(x[chunk], indices, roots, loopOrder, i ..< j,
+                                   atLayer + chunkLayer)
+            i = j
+  elif T is List:
+    const
+      totalChunks = maxChunksCount(T, x.maxLen)
+      firstChunkIndex = nextPow2(totalChunks.uint64)
+      chunkLayer = log2trunc(firstChunkIndex)
+    var i = slice.a
+    while i <= slice.b:
+      let
+        index = indexAt(i)
+        indexLayer = log2trunc(index)
+      if index == 1.GeneralizedIndex:
```

Comment:
Most of this logic is duplicated above and below. What are the differences?

Reply:
Indeed, I have not found an elegant way around the duplication.

Four of the SSZ types get hashed according to the same format, with a subtree for the actual data to the left (`2.GeneralizedIndex`), and some metadata to the right (`3.GeneralizedIndex`).

- `BitList` (aka variable-sized `BitArray`) puts the members (padded with zeroes until capacity) in a subtree to the left (`2.GeneralizedIndex`), and put the current length to the right (`3.GeneralizedIndex`). It works using `bitListHashTreeRoot`.
- `List` (aka variable-sized `array`) puts the members (padded with zeroes until capacity) in a subtree to the left (`2.GeneralizedIndex`), and put the current length to the right (`3.GeneralizedIndex`). It works using `chunkedHashTreeRoot`. Compared to `BitList` it also takes into account that members are more complex than individual bits, so the code is not exactly same as in `BitList`.
- `HashList` (aka variable-sized `array` with a cache for intermediate hashes) puts the members (padded with zeroes until capacity) in a subtree to the left (`2.GeneralizedIndex`), and put the current length to the right (`3.GeneralizedIndex`). It works using `hashTreeRootCached`. Also calls directly into `chunkedHashTreeRoot` for small member sizes, so not exactly same as `BitList` / `List`.
- `SingleMemberUnion` puts the value in a subtree to the left (`2.GeneralizedIndex`), and puts the selector / union kind to the right (`3.GeneralizedIndex`). It works by hardcoding the selector to `0` as that is the only one supported for regular `hash_tree_root` by this codebase as well.

The code duplication mainly resembles the pre-existing code duplication for the existing `hash_tree_root` function. Sadly, they are all slightly different. Maybe some `template` shenanigans could avoid _some_ of the duplication, but it does not seem to make it easier for actually understanding it. As we have "only" four similar things in a file that rarely has to be checked, I went with the current approach of making the individual implementations look similar (so that once one of them is understood, all of them are understood), and just keeping the tiny differences in each of the four copies.

One way to review these segments is to search the file for `3.GeneralizedIndex` and then to keep swapping back and forth with `cmd-G` / `cmd-shift-G`. Editors like VSCode / CLion show the search results roughly at the same location of the screen when jumping around like this, allowing a visual comparison.